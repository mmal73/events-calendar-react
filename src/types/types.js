export const types = {
    uiOpenModal: '[ui] Open modal',
    uiCloseModal: '[ui] Close modal',
    eventAddNew: '[event] Add new',
    eventSetActive: '[event] Set active',
    eventClearActive: '[event] Clear active',
    eventUpdated: '[event] Updated',
    eventDeleted: '[event] Deleted',
    calendarDateSelected: '[calendar] Date selected',
}