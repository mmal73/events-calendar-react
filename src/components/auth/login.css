.login-container {
    width: 100%;
    background: var(--primary);
    background: -webkit-linear-gradient(to right, var(--secundary), var(--primary));
    background: linear-gradient(to right, var(--secundary), var(--primary));
    display: flex;
    align-items: center;
    height: 100vh;
}

.forms-container {
    margin: 0 auto;
    border-radius: 5px;
    width: 90%;
    height: 100%;
}

.login-form-1 {
    background: rgba(255, 255, 255, .2);

}

.login-form-2 {
    background: rgba(255, 255, 255, .4);
}

.input-form {
    margin-bottom: .5rem;
}

.btn-submit {
    background: #ffffff;
    margin-top: 1rem;
    width: 50%;
    border-radius: 1rem;
    padding: 1.5%;
    border: none;
    cursor: pointer;
}