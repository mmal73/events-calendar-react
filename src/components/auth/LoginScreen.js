import React from 'react';
import './login.css';

export default function LoginScreen() {
    return (
        <div className="login-container">
            <div className="container">
                <div className="row forms-container">
                    <div className="col-md-6 login-form-1 mt-2 p-4">
                        <h3 className="text-white text-center">Ingreso</h3>
                        <form className="form-container">
                            <div className="form-group">
                                <input
                                    type="text"
                                    className="form-control input-form"
                                    placeholder="Correo"
                                />
                            </div>
                            <div className="form-group">
                                <input
                                    type="password"
                                    className="form-control input-form"
                                    placeholder="Contraseña"
                                />
                            </div>
                            <div className="form-group d-flex justify-content-center">
                                <input
                                    type="submit"
                                    className="btn-submit"
                                    value="Login"
                                />
                            </div>
                        </form>
                    </div>
                    <div className="col-md-6 login-form-2 mt-2 p-4">
                        <h3 className="text-white text-center">Registro</h3>
                        <form>
                            <div className="form-group">
                                <input
                                    type="text"
                                    className="form-control input-form"
                                    placeholder="Nombre"
                                />
                            </div>
                            <div className="form-group">
                                <input
                                    type="email"
                                    className="form-control input-form"
                                    placeholder="Correo"
                                />
                            </div>
                            <div className="form-group">
                                <input
                                    type="password"
                                    className="form-control input-form"
                                    placeholder="Contraseña"
                                />
                            </div>
                            <div className="form-group">
                                <input
                                    type="password"
                                    className="form-control input-form"
                                    placeholder="Repita la contraseña"
                                />
                            </div>
                            <div className="form-group d-flex justify-content-center">
                                <input
                                    type="submit"
                                    className="btn-submit"
                                    value="Crear cuenta" />
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        </div>
    )
}