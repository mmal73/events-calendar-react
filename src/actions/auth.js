import { fetchWithoutToken, fetchWithToken } from "../helpers/myFetch";
import { types } from "../types/types";
import Swal from "sweetalert2";
import { eventClear } from "./events";

const saveToken = ( token ) => {
    localStorage.setItem( 'token', token );
    localStorage.setItem( 'token-init-date', new Date().getTime() );
}

const showErrors = (message, errors) => {
    if( !message ){
        message = '';
        for (const field in errors) {
            if (errors.hasOwnProperty(field)) {
                const _error = errors[field].msg;
                message += _error;
                message += `<br>`;
            }
        }
    }
    return Swal.fire({
        icon: 'error',
        title: 'Oops...',
        html: message
    });
}

const login = ( user ) => ({
    type: types.authLogin,
    payload: user
});

const checkingFinish = () => ({
    type: types.authFinishCheking
});

const authLoginStart = ( email, password ) => {
    return async( dispatch ) => {
        const resp = await fetchWithoutToken('auth/login', { email, password }, 'POST');
        const data = await resp.json();
        if( data.status ){
            const { token, _id, name } = data;
            saveToken(token);
            dispatch( login({ _id, name }) )
        }else{
            showErrors( data.message, data.errors );
        }

    }
}

const startRegister = ( name, email, password ) => {
    return async( dispatch ) => {
        const resp = await fetchWithoutToken('auth/register', { name, email, password }, 'POST');
        const data = await resp.json();
        if( data.status ){
            const { token, _id } = data;
            saveToken(token);
            dispatch( login({ _id, name }) )
        }else{
            showErrors( data.message, data.errors );
        }
    };
}

const startChecking = () => {
    return async( dispatch ) => {
        const resp = await fetchWithToken('auth/renew-token');
        const data = await resp.json();
        if( data.status ){
            const { token, _id, name } = data;
            saveToken(token);
            dispatch( login({ _id, name }) )
        }else{
            dispatch( checkingFinish() );
        }
    }
}

const logout = () => ({
    type: types.authLogout
})
const startLogout = () => {
    return ( dispatch ) => {
        localStorage.removeItem('token');
        localStorage.removeItem('token-init-date');
        dispatch( eventClear() );
        dispatch( logout() );
    }
}

export{
    authLoginStart,
    startRegister,
    startChecking,
    startLogout,
    showErrors
}
