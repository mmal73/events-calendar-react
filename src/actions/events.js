import { types } from "../types/types"
import {showErrors} from './auth';
import { fetchWithToken } from "../helpers/myFetch";
import { prepareEvents } from "../helpers/prepareEvents";
export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
})
export const eventClearActive = (event) => ({
    type: types.eventSetActive
})
const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
})
export const eventStartAddNew = ( event ) => {
    return async( dispatch, getState ) => {
        const resp = await fetchWithToken('events', event, 'POST');
        const data = await resp.json();
        const { _id, name } = getState().auth;
        if( data.status ){
            event.id = data.event.id;
            event.user = {
                _id,
                name
            }
            dispatch( eventAddNew( event ) );
        }else{
            showErrors( data.message, data.errors );
        }
    }
}
const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});
export const startEventUpdated = ( event ) => {
    return async( dispatch ) => {
        try {
            const resp = await fetchWithToken(`events/${event.id}`, event, 'PUT');
            const data = await resp.json();
            if( data.status ){
                dispatch( eventUpdated( event ) );
            }else{
                showErrors( data.message, data.errors );
            }
        } catch ( error ) {
            console.log(error)
        }
    }
};
export const eventDeleted = () => ({
    type: types.eventDeleted
});
export const eventStartDelete = () => {
    return async( dispatch, getState ) => {
        const { id } = getState().calendar.eventActive;
        try {
            const resp = await fetchWithToken(`events/${id}`, {}, 'DELETE');
            const data = await resp.json();
            if( data.status ){
                dispatch( eventDeleted() );
            }else{
                showErrors( data.message, data.errors );
            }
        } catch ( error ) {
            console.log(error)
        }
    }
};
export const calendarDateSelected = (event) => ({
    type: types.calendarDateSelected,
    payload: event
});
const eventLoaded = ( events ) => ({
    type: types.eventLoaded,
    payload: events
})
export const eventStartLoaded = () => {
    return async( dispatch ) => {
        try {
            const resp = await fetchWithToken('events');
            const data = await resp.json();
            if( data.status ){
                const preparedEvents = prepareEvents( data.events );
                dispatch( eventLoaded( preparedEvents ) );
            }else{
                showErrors( data.message, data.errors );
            }
        } catch (error) {
            console.log(error);
        }
    }
}
