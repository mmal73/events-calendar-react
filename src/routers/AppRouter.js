import React, { useEffect } from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Redirect
} from "react-router-dom";
import { useDispatch, useSelector } from 'react-redux';

import LoginScreen from '../components/auth/LoginScreen';
import CalendarScreen from '../components/calendar/CalendarScreen';
import { startChecking } from '../actions/auth';
import PrivateRoute from './PrivateRoute';
import PublicRoute from './PublicRoute';

export default function AppRouter() {

    const dispatch = useDispatch();
    const { checking, _id } = useSelector( state => state.auth);
    
    const authenticated = !!_id;

    useEffect(() => {
        dispatch( startChecking() );
    }, [dispatch])

    if ( checking ) {
        return ( <h3>Checking...</h3> )
    }
    return (
        <Router basename={process.env.PUBLIC_URL}>
            <div>
                <Switch>
                    <PublicRoute exact path={`${process.env.PUBLIC_URL}/login`} component={LoginScreen} isAuthenticated={authenticated} />
                    <PrivateRoute exact path={`${process.env.PUBLIC_URL}/`} component={CalendarScreen} isAuthenticated={authenticated} />
                    <Redirect to="/"/>
                </Switch>
            </div>
        </Router>
    )
}
