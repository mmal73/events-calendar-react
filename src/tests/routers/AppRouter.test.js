import React from 'react';
import { Provider } from "react-redux";
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import AppRouter from '../../routers/AppRouter';
import { startChecking } from '../../actions/auth';
import { types } from '../../types/types';

jest.mock('sweetalert2', () => ({
    fire: jest.fn()
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );
describe('Tests in AppRouter', () => {
    
    test('correct authentication checking', async() => {
        const initState = {
            auth: {
                checking: true
            }
        };

        const store = mockStore( initState );
        await store.dispatch( startChecking() );
        expect( store.getActions()[0] ).toEqual({ type: types.authFinishCheking });

        const wrapper = mount(
            <Provider store={ store }>
                <AppRouter/>
            </Provider>
        );
        expect( wrapper ).toMatchSnapshot();
    }); 
    
    test('should show login screen', async() => {
        const initState = {
            auth: {
                checking: false,
                _id: null
            }
        };

        const store = mockStore( initState );
        await store.dispatch( startChecking() );
        expect( store.getActions()[0] ).toEqual({ type: types.authFinishCheking });

        const wrapper = mount(
            <Provider store={ store }>
                <AppRouter/>
            </Provider>
        );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( '.login-container' ).exists() ).toBe( true );
    });
    test('should show calendar screen', async() => {
        const initState = {
            auth:{
                checking: false,
                _id: '123',
                name: 'userTest'
            },
            calendar:{
                events: []
            },
            ui:{
                modalOpen: false
            }
        };
        
        const store = mockStore( initState );
        await store.dispatch( startChecking() );
        expect( store.getActions()[0] ).toEqual({ type: types.authFinishCheking });

        const wrapper = mount(
            <Provider store={ store }>
                <AppRouter/>
            </Provider>
        );
        expect( wrapper.find( '.btn-primary' ).exists() ).toBe( true );
    });
})
