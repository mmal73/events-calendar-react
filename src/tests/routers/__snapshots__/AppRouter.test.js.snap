// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in AppRouter correct authentication checking 1`] = `
<h3>
  Checking...
</h3>
`;

exports[`Tests in AppRouter should show login screen 1`] = `
<div>
  <div
    className="login-container"
  >
    <div
      className="container"
    >
      <div
        className="row forms-container"
      >
        <div
          className="col-md-6 login-form-1 mt-2 p-4"
        >
          <h3
            className="text-white text-center"
          >
            Ingreso
          </h3>
          <form
            className="form-container"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <input
                autoComplete="off"
                className="form-control input-form"
                name="lEmail"
                onChange={[Function]}
                placeholder="Correo"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                autoComplete="off"
                className="form-control input-form"
                name="lPassword"
                onChange={[Function]}
                placeholder="Contraseña"
                type="password"
                value=""
              />
            </div>
            <div
              className="form-group d-flex justify-content-center"
            >
              <input
                autoComplete="off"
                className="btn-submit"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
        <div
          className="col-md-6 login-form-2 mt-2 p-4"
        >
          <h3
            className="text-white text-center"
          >
            Registro
          </h3>
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <input
                autoComplete="off"
                className="form-control input-form"
                name="rName"
                onChange={[Function]}
                placeholder="Nombre"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                autoComplete="off"
                className="form-control input-form"
                name="rEmail"
                onChange={[Function]}
                placeholder="Correo"
                type="email"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                autoComplete="off"
                className="form-control input-form"
                name="rPassword1"
                onChange={[Function]}
                placeholder="Contraseña"
                type="password"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                autoComplete="off"
                className="form-control input-form"
                name="rPassword2"
                onChange={[Function]}
                placeholder="Repita la contraseña"
                type="password"
                value=""
              />
            </div>
            <div
              className="form-group d-flex justify-content-center"
            >
              <input
                autoComplete="off"
                className="btn-submit"
                type="submit"
                value="Crear cuenta"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
